utils/modules/edge_utils.py

```diff
@@ -81,3 +130,135 @@ def faces_angle(normals, edges, pols):
             ang = 2*pi
         angles.append(ang)
     return angles
+
+def edges_normals_full(vertices, edges, faces):
+    '''Average of normals of the faces that share the edge (only the two first)'''
+    bm = bmesh_from_pydata(vertices, edges, faces, normal_update=True)
+    normals = [tuple(face.normal) for face in bm.faces]
+    bm.free()
+    vals = edges_normal(vertices, normals, edges, faces)
+
+    return vals
+
+def edges_normal(vertices, normals, edges, pols):
+    '''Average of normals of the faces that share the edge (only the two first)'''
+    e_sorted = [sorted(e) for e in edges]
+    ad_faces = [[] for e in edges]
+    for idp, pol in enumerate(pols):
+        for edge in zip(pol, pol[1:] + [pol[0]]):
+            e_s = sorted(edge)
+            if e_s in e_sorted:
+                idx = e_sorted.index(e_s)
+                ad_faces[idx].append(idp)
+    result = []
+
+    for edg_f, edg in zip(ad_faces, edges):
+        if len(edg_f) > 1:
+            edge_normal = ((Vector(normals[edg_f[0]])+Vector(normals[edg_f[1]]))/2).normalized()
+        elif len(edg_f) == 1:
+            edge_normal = Vector(normals[edg_f[0]])
+        else:
+            rot_mat = Matrix.Rotation(pi/2, 4, "Z")
+            direc = (Vector(vertices[edg[1]]) - Vector(vertices[edg[0]])).normalized()
+            edge_normal = direc @ rot_mat
+        result.append(tuple(edge_normal))
+
+    return result
+
+
+def edges_vertices(vertices, edges):
+    '''Explode edges'''
+    verts = [[vertices[c] for c in e] for e in edges]
+    eds = [[[0, 1]] for e in edges]
+    vals = [verts, eds]
+    return vals
+
+def edge_is_filter(vertices, edges, faces, mode):
+    '''acces to mesh_filter to get differnt bmesh edges filters'''
+    mode = mode.split(' ')[1]
+    bm = bmesh_from_pydata(vertices, edges, faces, normal_update=True)
+    good, bad, mask = Edges.process(bm, mode, edges)
+    bm.free()
+    return mask, good, bad
+
+
+def edges_shell_factor(vertices, edges, faces):
+    '''Average of vertex shell_factor'''
+    v_shell = vertex_shell_factor(vertices, edges, faces)
+    vals = [(v_shell[e[0]] + v_shell[e[1]])/2 for e in edges]
+    return vals
+
+def edges_center(vertices, edges):
+    vals = [tuple((Vector(vertices[e[0]])+Vector(vertices[e[1]]))/2) for e in edges]
+    return vals
+
+def edges_origin(vertices, edges):
+    vals = [vertices[e[0]] for e in edges]
+    return vals
+
+def edges_end(vertices, edges):
+    vals = [vertices[e[1]] for e in edges]
+    return vals
+
+def edges_inverted(vertices, edges):
+    vals = [[e[1], e[0]] for e in edges]
+    return vals
```

Comment:
```Python
def edges_inverted(vertices, edges):
    return [[vertices[i1], vertices[i2]] for i1, i2 in edges]

Reply:
Actually should be:
```
def edges_inverted(edges):
    '''
    edges: list with edges [[int, int], [int,int]...]
    outputs the same list with inverted index [[0, 1],[1, 2]] ---> [[1, 0], [2, 1]]
    '''
    return [[e[1], e[0]] for e in edges]
```
    `

Reply:
Then socket type should be fixed.
![2020-01-19_07-34-45](https://user-images.githubusercontent.com/28003269/72674282-4ffd7880-3a8e-11ea-950f-1c6cbe0ea131.png)